### DISCUSSION TITLE
Triggering onNodesChange when node data updated

### DISCUSSION THREAD

**Discussion by ashwinsr** [xyflow/xyflow#2090] [Q&A] [answered]
It looks like updating the data field for a node will cause the node to re-render correctly but does not trigger onNodesChange. Is this the expected behavior? Is there any way to trigger onNodesChange when a node's data field is updated?

**Comment by bcakmakoglu** [comment, marked as answer]
That should be expected behavior.
`onNodesChange` is _not_ triggered when _any_ property in a node changes.
Rather it is triggered to reflect specific changes which you can either handle manually or use the `applyNodeChanges` utility to apply the default change-handlers.

`onNodesChange` will provide an Array of `NodeChange` as it's first argument (not an Array of `Node`).
You can find the types [here](https://github.com/wbkd/react-flow/blob/main/src/types/changes.ts)

In your case I would assume you would have to handle the data change yourself.
You could add your own callback in the data fields and call that from your custom node, when the data field changes and you want to "emit" that change to your parent component.

Something like this I guess
```jsx
// callback that receives the data change 
const onDataChange = useCallback((id, data) => {
	console.log(`data change in node ${id}`, data)
}, [])

const initialNodes = [
	id: '1',
	type: 'custom',
	position: { x: 0, y: 0, },
	data: {
		label: 'Node 1',
		myData: [],
		onDataChange, // pass the cb to your node
	}
]

// ... the rest of your jsx

**Reply by ashwinsr** [reply]
Thank you for the very speedy reply!

**Reply by bcakmakoglu** [reply]
Anytime, glad to help 😄
